client/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -0,0 +1,468 @@
+                    //mark host alive and don't retry, as the error should be a request problem
+                    onSuccess(host);
+                    throw lastSeenException;
+            }
+        }
+        //we get here only when we tried all nodes and they all failed
+        assert lastSeenException != null;
+        throw lastSeenException;
+    }
+
+    private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
+        Objects.requireNonNull(requestHeaders, "request headers must not be null");
+        for (Header defaultHeader : defaultHeaders) {
+            httpRequest.setHeader(defaultHeader);
+        }
+        for (Header requestHeader : requestHeaders) {
+            Objects.requireNonNull(requestHeader, "request header must not be null");
+            httpRequest.setHeader(requestHeader);
+        }
+    }
+
+    /**
+     * Returns an iterator of hosts to be used for a request call.
+     * Ideally, the first host is retrieved from the iterator and used successfully for the request.
+     * Otherwise, after each failure the next host should be retrieved from the iterator so that the request can be retried till
+     * the iterator is exhausted. The maximum total of attempts is equal to the number of hosts that are available in the iterator.
+     * The iterator returned will never be empty, rather an {@link IllegalStateException} in case there are no hosts.
+     * In case there are no healthy hosts available, or dead ones to be be retried, one dead host gets returned.
+     */
+    private Iterable<HttpHost> nextHost() {
+        Set<HttpHost> filteredHosts = new HashSet<>(hosts);
+        for (Map.Entry<HttpHost, DeadHostState> entry : blacklist.entrySet()) {
+            if (System.nanoTime() - entry.getValue().getDeadUntilNanos() < 0) {
+                filteredHosts.remove(entry.getKey());
+            }
+        }
+
+        if (filteredHosts.isEmpty()) {
+            //last resort: if there are no good hosts to use, return a single dead one, the one that's closest to being retried
+            List<Map.Entry<HttpHost, DeadHostState>> sortedHosts = new ArrayList<>(blacklist.entrySet());
+            Collections.sort(sortedHosts, new Comparator<Map.Entry<HttpHost, DeadHostState>>() {
+                @Override
+                public int compare(Map.Entry<HttpHost, DeadHostState> o1, Map.Entry<HttpHost, DeadHostState> o2) {
+                    return Long.compare(o1.getValue().getDeadUntilNanos(), o2.getValue().getDeadUntilNanos());
+                }
+            });
+            HttpHost deadHost = sortedHosts.get(0).getKey();
+            logger.trace("resurrecting host [" + deadHost + "]");
+            return Collections.singleton(deadHost);
+        }
+
+        List<HttpHost> rotatedHosts = new ArrayList<>(filteredHosts);
+        //TODO is it possible to make this O(1)? (rotate is O(n))
+        Collections.rotate(rotatedHosts, rotatedHosts.size() - lastHostIndex.getAndIncrement());
```

Comment:
rotate copies data, I think you don't need to do this if you use something like RotatedList we have in core... maybe it's ok to copy it for this purpose? But I think we can just stick with the collection one we have.  if you would not use a set it would be cheaper